File: frontend/src/api/modules/request-capture.ts
import type { RequestOptions } from '../request'
import { API_BASE_URL } from '../constants'
import request from '../request'

export interface CaptureConfig {
  enabled: boolean
  quotaMib: number
  retentionDays: number
}
export interface CaptureTask {
  id: string
  scope: 'key' | 'account' | 'group'
  targetId: string
  includeMedia: boolean
  startedAt: string
  expiresAt: string
  status: 'running' | 'stopped' | 'expired' | 'interrupted'
}
export interface CaptureRecord {
  id: string
  taskId: string
  requestId: string
  bytes: number
  incomplete: boolean
  createdAt: string
}
export interface CaptureStatus {
  config: CaptureConfig
  instanceId: string
  tasks: CaptureTask[]
  records: CaptureRecord[]
  skipped: number
  activeSessions: number
  bufferedBytes: number
  storageFault: boolean
}
export interface CapturePage { text: string, nextOffset: number | null }
export function getRequestCaptures(options: RequestOptions = {}) {
  return request<CaptureStatus>({ url: '/api/admin/request-captures', method: 'GET', ...options })
}
export function configureRequestCaptures(data: CaptureConfig) {
  return request<void>({ url: '/api/admin/request-captures/config', method: 'POST', data })
}
export function createRequestCapture(data: Pick<CaptureTask, 'scope' | 'targetId' | 'includeMedia'> & { minutes: number }) {
  return request<CaptureTask>({ url: '/api/admin/request-captures', method: 'POST', data })
}
export function stopRequestCapture(id: string) {
  return request<void>({ url: `/api/admin/request-captures/${encodeURIComponent(id)}/stop`, method: 'POST' })
}
export function deleteRequestCapture(id: string) {
  return request<void>({ url: `/api/admin/request-captures/${encodeURIComponent(id)}`, method: 'DELETE' })
}
export function readRequestCapture(id: string, offset: number, options: RequestOptions = {}) {
  return request<CapturePage>({ url: `/api/admin/request-captures/records/${encodeURIComponent(id)}`, method: 'GET', params: { offset }, ...options })
}
export function captureExportUrl(kind: 'task' | 'record', id: string) {
  const segment = kind === 'record' ? 'records/' : ''
  return `${API_BASE_URL}/api/admin/request-captures/${segment}${encodeURIComponent(id)}/export`
}

File: backend/crates/gateway-admin/src/model/request_capture.rs
//! Administrator-only, scoped and time-limited error capture.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::AdminError;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RequestCaptureConfig {
    pub enabled: bool,
    pub quota_mib: u32,
    pub retention_days: u16,
}

impl Default for RequestCaptureConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            quota_mib: 1024,
            retention_days: 7,
        }
    }
}

impl RequestCaptureConfig {
    pub fn validate(&self) -> Result<(), AdminError> {
        if !(1..=102_400).contains(&self.quota_mib) || !(1..=30).contains(&self.retention_days) {
            return Err(AdminError::invalid("采集配额或保留天数超出范围"));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CaptureScope {
    Key,
    Account,
    Group,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CreateCaptureTask {
    pub scope: CaptureScope,
    pub target_id: String,
    pub minutes: u16,
    #[serde(default)]
    pub include_media: bool,
}

impl CreateCaptureTask {
    pub fn validate(&self) -> Result<(), AdminError> {
        if !(1..=1440).contains(&self.minutes)
            || self.target_id.is_empty()
            || self.target_id.len() > 128
            || self.target_id.chars().any(char::is_control)
        {
            return Err(AdminError::invalid("采集范围或时长不合法"));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureTask {
    pub id: String,
    pub scope: CaptureScope,
    pub target_id: String,
    pub include_media: bool,
    pub started_at: DateTime<Utc>,
    pub expires_at: DateTime<Utc>,
    pub status: CaptureTaskStatus,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CaptureTaskStatus {
    Running,
    Stopped,
    Expired,
    Interrupted,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureRecord {
    pub id: String,
    pub task_id: String,
    pub request_id: String,
    pub bytes: u64,
    pub incomplete: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RequestCaptureStatus {
    pub config: RequestCaptureConfig,
    pub instance_id: String,
    pub tasks: Vec<CaptureTask>,
    pub records: Vec<CaptureRecord>,
    pub skipped: u64,
    pub active_sessions: usize,
    pub buffered_bytes: usize,
    pub storage_fault: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CapturePage {
    pub text: String,
    pub next_offset: Option<u64>,
}
